### PR TITLE
fix(cache): Add cache warmup for non-legacy charts

### PR DESCRIPTION
**Pull request by john-bodley**
### SUMMARY

This PR is a rehash of https://github.com/apache/superset/pull/23012 which was later reverted by https://github.com/apache/superset/pull/23579 in combination with https://github.com/apache/superset/pull/23569.

The logic defined here is the inverse of https://github.com/apache/superset/pull/23012, i.e., rather than trying to identify if a chart is a non-legacy chart (via the presence of the query-context—which was problematic), instead we check if the chart is a supported legacy chart/viz type.

### TESTING INSTRUCTIONS

Added unit tests.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

cc: @alexryndin @dheeraj-jaiswal-lowes

### CODE REVIEW

File: superset/charts/commands/warm_up_cache.py [hunk 1]
```diff
@@ -43,31 +45,51 @@ def __init__(
     def run(self) -> dict[str, Any]:
         self.validate()
         chart: Slice = self._chart_or_id  # type: ignore
+
         try:
             form_data = get_form_data(chart.id, use_slice_data=True)[0]
-            if self._dashboard_id:
```
Comment (john-bodley) [inline superset/charts/commands/warm_up_cache.py:48]: This logic (for the most part) has just been indented.

File: superset/views/core.py [hunk 2]
```diff
@@ -825,43 +827,15 @@ def warm_up_cache(  # pylint: disable=too-many-locals,no-self-use
                 .all()
             )
 
-        result = []
```
Comment (john-bodley) [inline superset/views/core.py:828]: Adhering to the DRY principle.

File: tests/integration_tests/charts/api_tests.py [hunk 3]
```diff
@@ -14,37 +14,41 @@
 # KIND, either express or implied.  See the License for the
 # specific language governing permissions and limitations
 # under the License.
-# isort:skip_file
```
Comment (john-bodley) [inline tests/integration_tests/charts/api_tests.py:17]: I have no idea why this was here.

File: tests/integration_tests/charts/api_tests.py [hunk 4]
```diff
@@ -1710,12 +1714,16 @@ def test_gets_owned_created_favorited_by_me_filter(self):
         assert data["result"][0]["slice_name"] == "name0"
         assert data["result"][0]["datasource_id"] == 1
 
-    @pytest.mark.usefixtures(
-        "load_energy_table_with_slice", "load_birth_names_dashboard_with_slices"
+    @parameterized.expand(
```
Comment (john-bodley) [inline tests/integration_tests/charts/api_tests.py:1717]: Ensure we can successfully warm up legacy and non-legacy charts.

File: tests/integration_tests/core_tests.py [hunk 5]
```diff
@@ -14,72 +14,67 @@
 # KIND, either express or implied.  See the License for the
 # specific language governing permissions and limitations
 # under the License.
-# isort:skip_file
```
Comment (john-bodley) [inline tests/integration_tests/core_tests.py:17]: I have no idea why this was here.

File: tests/integration_tests/charts/api_tests.py [hunk 6]
```diff
@@ -1710,12 +1714,16 @@ def test_gets_owned_created_favorited_by_me_filter(self):
         assert data["result"][0]["slice_name"] == "name0"
         assert data["result"][0]["datasource_id"] == 1
 
-    @pytest.mark.usefixtures(
-        "load_energy_table_with_slice", "load_birth_names_dashboard_with_slices"
```
Comment (john-bodley) [inline tests/integration_tests/charts/api_tests.py:1714]: The `load_energy_table_with_slice` fixture is not required.

File: tests/integration_tests/core_tests.py [hunk 7]
```diff
@@ -386,9 +381,7 @@ def test_databaseview_edit(self, username="admin"):
         database.allow_run_async = False
         db.session.commit()
 
-    @pytest.mark.usefixtures(
-        "load_energy_table_with_slice", "load_birth_names_dashboard_with_slices"
```
Comment (john-bodley) [inline tests/integration_tests/core_tests.py:390]: The `load_energy_table_with_slice` fixture is not required.

File: superset/charts/commands/warm_up_cache.py [hunk 8]
```diff
@@ -43,31 +48,51 @@ def __init__(
...
-            g.form_data = form_data
-            payload = obj.get_payload()
-            delattr(g, "form_data")
-            error = payload["errors"] or None
-            status = payload["status"]
+
+            if form_data.get("viz_type") in viz_types:
+                # Legacy visualizations.
+                if not chart.datasource:
+                    raise ChartInvalidError("Chart's datasource does not exist")
```
Comment (john-bodley) [inline superset/charts/commands/warm_up_cache.py:58]: Raising a `ChartInvalidError` seems more appropriate than raising an `Exception`.

File: superset/views/core.py [hunk 9]
```diff
@@ -769,7 +770,8 @@ def save_or_overwrite_slice(
     @api
     @has_access_api
     @expose("/warm_up_cache/", methods=("GET",))
-    def warm_up_cache(  # pylint: disable=too-many-locals,no-self-use
+    @deprecated(new_target="api/v1/chart/warm_up_cache/")
```
Comment (john-bodley) [inline superset/views/core.py:773]: @michael-s-molina I gather it's now too late to remove this endpoint in 3.0 given there's a new target.

File: tests/integration_tests/charts/commands_tests.py [hunk 10]
```diff
@@ -96,52 +96,7 @@ def test_export_chart_command(self, mock_g):
             "dataset_uuid": str(example_chart.table.uuid),
             "uuid": str(example_chart.uuid),
             "version": "1.0.0",
-        }
```
Comment (john-bodley) [inline tests/integration_tests/charts/commands_tests.py:99]: Adhering to the logic in https://github.com/apache/superset/pull/23012.

File: superset/charts/commands/export.py [hunk 11]
```diff
@@ -34,7 +34,7 @@
 
 
 # keys present in the standard export that are not needed
-REMOVE_KEYS = ["datasource_type", "datasource_name", "query_context", "url_params"]
+REMOVE_KEYS = ["datasource_type", "datasource_name", "url_params"]
```
Comment (michael-s-molina) [inline superset/charts/commands/export.py:37]: Why did you add the `query_context` to the payload of exported charts?

File: superset/charts/commands/warm_up_cache.py [hunk 12]
```diff
@@ -43,31 +48,51 @@ def __init__(
...
-
-            # pylint: disable=assigning-non-slot
-            g.form_data = form_data
-            payload = obj.get_payload()
-            delattr(g, "form_data")
-            error = payload["errors"] or None
-            status = payload["status"]
+
+            if form_data.get("viz_type") in viz_types:
+                # Legacy visualizations.
```
Comment (michael-s-molina) [inline superset/charts/commands/warm_up_cache.py:56]: The problem with this approach is that we have non-legacy charts that are also present in `viz_types` under the same key like `table` and `big_number`.

**Review by michael-s-molina — COMMENTED** [review]
Thank you for the PR @john-bodley. I left some first-pass comments.

File: superset/charts/commands/export.py (hunk 11 above)
Comment (john-bodley) [reply superset/charts/commands/export.py:37]: @michael-s-molina this logic was extracted from https://github.com/apache/superset/pull/23012.

File: superset/charts/commands/warm_up_cache.py (hunk 12 above)
Comment (john-bodley) [reply superset/charts/commands/warm_up_cache.py:56]: @michael-s-molina that may explain why testing with the `table` viz type was problematic.

Should the `table`, `big_number`, etc. viz types be removed from `superset.viz`? Alternatively if they’re still required, i.e., we simultaneously support both legacy and non-legacy versions, how does Superset differentiate between the two? Is it the presence of the query context?

Comment (michael-s-molina) [reply superset/charts/commands/warm_up_cache.py:56]: > Should the table, big_number, etc. viz types be removed from superset.viz? 

Yes. Let's remove them.

File: superset/views/core.py (hunk 9 above)
Comment (dpgaspar) [reply superset/views/core.py:773]: unfortunately yes

Comment (michael-s-molina) [reply superset/views/core.py:773]: Let's remove it in 4.0 given that the breaking window is closed.

File: superset/charts/commands/warm_up_cache.py (hunk 12 above)
Comment (john-bodley) [reply superset/charts/commands/warm_up_cache.py:56]: See https://github.com/apache/superset/pull/24694 which is now a precursor to this PR.

**Comment by john-bodley** [comment]
@michael-s-molina this PR is now ready for re-review.

**Review by michael-s-molina — APPROVED** [review]
LGTM. I left a non-blocking comment about the status when there's an error.

File: superset/charts/commands/warm_up_cache.py [hunk 13]
```diff
@@ -43,31 +48,51 @@ def __init__(
...
+                # Report the first error.
+                for query in payload["queries"]:
+                    error = query["error"]
+                    status = query["status"]
+
+                    if error is not None:
+                        break
         except Exception as ex:  # pylint: disable=broad-except
             error = error_msg_from_exception(ex)
             status = None
```
[the remark this answers is not in the document]
Comment (john-bodley) [reply superset/charts/commands/warm_up_cache.py:98]: It seems not. This is the same logic as previously.

**Comment by stepcheunghk** [comment]
@john-bodley Thank you so much. This one is very helpful!

**Comment by stepcheunghk** [comment]
Hi @john-bodley, we have found something funny. We noticed that some charts have been cached by cache warmup job, which is good. But if we click the dashboard, it seems those charts in the dashboard are not using the warmup cache. Do you have any idea? Thank you.

**Comment by dheeraj281** [comment]
Hi @stepcheunghk , Are there any native filters or other filters applied to the chart as part of dashboard?

**Comment by stepcheunghk** [comment]
Hi @dheeraj281 , you are right. Definitely there are some filters applied to the chart/ dashboard. What should I do now? Thanks man.

**Comment by dheeraj281** [comment]
@stepcheunghk Yes, thats the reason.. It refreshes the chart objects related to the dashboard. When native filter is applied on the dashboard, its a different query now that is why. If you dont keep any filter with default, it should work as expected.

**Comment by stepcheunghk** [comment]
@dheeraj281 I see. Thanks for your prompt reply and clear explanation. 🙂

**Comment by oriabyi** [comment]
Hi @dheeraj281 @stepcheunghk 
if I have a dashboard with a native filter with a default value and also in it, I have charts with filters inside (with clauses where/having), how should I do cache_warmup so it would refresh my dashboard with default values and all charts inside it with their own filters?

**Comment by dheeraj281** [comment]
Hi @oriabyi I am not aware of any such functionality right now. As discussed above It refreshes the chart objects related to the dashboard. it basically loads the chart url with force=true. One way of refreshing dashboard could be a separate selenium script that loads the dashboard url with url param "force=true"

Even we had this requirement of refreshing a dashboard with a native filter with a default value and also in it. To handle that in superset for our internal use, I have changed the cache-refresh functionality to work from the backend instead of using selenium and  chart url approach. In backend it will identify charts and all the associated dashboard level filters of it and runs the exact same query which is their at dashboard level using chartDataCommand. If many users wants this, I can create a PR for the same. 
cc: @rusackas @villebro

**Comment by oriabyi** [comment]
@dheeraj281, thank you, it helped a lot, truly appreciate it.
